application/models/Template.php

```diff
@@ -414,23 +411,31 @@ public static function getTemplatesFileFolder($sTemplateName = "")
     }
 
     /**
-     * Returns an array of all available template names - does a basic check if the template might be valid
-     *
-     * TODO: replace the calls to that function by a data provider based on search
-     *
-     * @return array
+     * Returns an array of all available template names - check if template exist
+     * key is template name, value is template folder
+     * @return string|]
      */
     public static function getTemplateList()
     {
-
-
-        $aTemplateList = array();
-
-        $oTemplateList = TemplateConfiguration::model()->search();
-        $oTemplateList->setPagination(false);
-
-        foreach ($oTemplateList->getData() as $oTemplate) {
-            $aTemplateList[$oTemplate->template_name] =  (self::isStandardTemplate($oTemplate->template_name)) ?  Yii::app()->getConfig("standardthemerootdir") . DIRECTORY_SEPARATOR . $oTemplate->template->folder : Yii::app()->getConfig("userthemerootdir") . DIRECTORY_SEPARATOR . $oTemplate->template->folder;
+        static $aTemplateList =  null;
+        if (!is_null($aTemplateList)) {
+            return $aTemplateList;
+        }
+        $aTemplateList = [];
+        /* Get the template name by TemplateConfiguration and fiolder by template , no need other data */
+        $criteria = new CDBCriteria();
+        $criteria->select = 'template_name';
+        $criteria->with = ['template' => ['select' => 'folder']];
+        $criteria->order = 'template_name';
+        $oTemplateList = TemplateConfiguration::model()->with('template')->findAll($criteria);
+        $aTemplateInStandard = SurveyThemeHelper::getTemplateInStandard();
+        $aTemplateInUpload = SurveyThemeHelper::getTemplateInUpload();
+        foreach ($oTemplateList as $oTemplate) {
```

Comment:
We need to check if directory of the final template is still OK,

I don't understand your proposition here … what code do you want ?

Reply:
> I don't understand why you are looping through $oTemplateList.

It mut be installed i think. Let me check

Reply:
> I don't understand why you are looping through $oTemplateList.

Else : here, it's because i take previous code using `TemplateConfiguration::model()->search()` , i move it to `TemplateConfiguration::model()->find`

Less update, but you'r right : we can surely use array_merge + array_intersect

Reply:
> ` $aTemplateList = array_merge($aTemplateInStandard, $aTemplateInUpload);`

We need
```
array
(
    '  extends_Extends_Vanilla  ' =>
'/media/shnoulle/data/webdev/ls5/upload/themes/survey/ 
extends_extends_vanilla  '
    '  extends_extends_vanilla  ' =>
'/media/shnoulle/data/webdev/ls5/upload/themes/survey/ 
extends_extends_vanilla  '
    'bootswatch' =>
'/media/shnoulle/data/webdev/ls5/themes/survey/bootswatch'
    'extends_vanilla' =>
'/media/shnoulle/data/webdev/ls5/upload/themes/survey/extends_vanilla'
    'fruity' => '/media/shnoulle/data/webdev/ls5/themes/survey/fruity'
    'vanilla' => '/media/shnoulle/data/webdev/ls5/themes/survey/vanilla'
)
```

And need it set in Db too (must have a directory + a DB option)

Maybe can do more array update to get it. Unsure if it's really more speed